chore: update examples to use `uv`

hasansezertasan opened the pull request:
Updating the examples to use `uv`.

> There are some additional changes in the examples that are not mainly related to the issue, such as using the same import statement along with the other examples, using the same database connection string, using `README.md` instead of `README.rst`, etc.

fixes #2621 and #2609

<!--
Ensure each step in the contributing guide is complete, especially the following:

- Add an entry in changelog.rst summarizing the change and linking to the issue.
-->

---

Progress Tracking:

- [x] `auth`:
- [x] `auth-flask-login`
- [x] `azure-blob-storage`:
    - [x] Use `testcontainers` to fire up an Azure Blob Storage container.
- [x] `babel`: Requires `sqlalchemy-with-utils` optional dependency to run.
- [x] `bootstrap4`: Requires `sqlalchemy-with-utils` optional dependency to run.
- [x] `csp-nonce`
- [x] `custom-layout`
- [x] `datetime-timezone`
- [x] `forms-files-images`
    - [x] Use `testcontainers` to fire up an Redis container.
    - [x] Move RedisCLI to it's own example.
- [x] `geo-alchemy`:
- [x] `host-matching`: How to produce the environment (IDK much about this `host="subdomain.localhost"` stuff)?
- [x] `methodview`
- [x] `multiple-admin-instances`
- [x] `peewee`: Requires `sqlalchemy` optional dependency to run.
- [x] `pymongo`
    - [x] Use `testcontainers` to fire up an MongoDB container.
- [x] `s3`:
- [x] `simple`
- [x] `sqla`:
- [x] `sqla-association_proxy`
- [x] `sqla-custom-inline-forms`
- [x] `tinymongo`: Has some issues and requires `sqlalchemy` optional dependency to run.

samuelhwilliams commented on examples/babel/README.md:
````diff
@@ -0,0 +1,26 @@
+
+> This example uses [`uv`](https://docs.astral.sh/uv/) to manage its dependencies and developer environment.
+
+To install the minimum version of Python this example supports, create your virtual environment, and install the required dependencies, run:
+
+```bash
+uv sync
+```
+
+Run the application:
+
+```bash
+uv run main.py
+```
````
```suggestion
Run the example using `uv`, which will manage the environment and dependencies automatically:

```bash
uv run main.py
```
```

hasansezertasan commented:
> Nice, tested the few you've done so far and this feels good - well done 🙇 

Thank you sir 🙏.

> I think we could also wrap into this standardising the 'entrypoint' file as 'main.py'? Some seem to use `app.py` (eg host-matching).

I agree with you, I haven't touch the other examples yet but I'll do it ASAP.

> Knowing that all the examples can just be run with `uv run main.py` feels nice

I see you, that's too convenient!

Thank you for thw great review 🫶. My laptop just died. I'll push further changes when I am able to 😁.

samuelhwilliams commented on .gitignore:
```diff
@@ -29,3 +29,4 @@ env
 .tox/
 .env
 doc/_build
+examples/**/uv.lock
```
Lock files should generally be committed to help with reproducible environments :)

samuelhwilliams commented on examples/auth-flask-login/README.md:
````diff
@@ -0,0 +1,28 @@
+uv run main.py
+```
+
+The first time you run this example, a sample sqlite database gets populated automatically. To suppress this behaviour, comment the following lines in main.py:
+
+```python
+if not os.path.exists(database_path):
+    with app.app_context():
+        build_sample_db()
+```
````
Is this advice that we definitely want to include? The examples don't work if you don't have a DB set up - I think this feels like a not useful recommendation?

hasansezertasan replied:
I don't mind removing this advice, some other examples also do have this one. Should I just go and remove them all @samuelhwilliams @ElLorans?

samuelhwilliams commented on examples/azure-blob-storage/main.py:
```diff
@@ -0,0 +1,31 @@
+
+
+if __name__ == "__main__":
+    with AzuriteContainer() as azurite_container:
```
Nice change - this feels helpful to get the example up and running quickly and reliably 👌

samuelhwilliams commented on examples/datetime-timezone/pyproject.toml:
```diff
@@ -0,0 +1,12 @@
+readme = "README.md"
+requires-python = ">=3.9"
+dependencies = [
+    "flask-admin[sqlalchemy]",
```
This example doesn't run yet:

```
flask-admin/examples/datetime-timezone on  chore/uv-migration/examples is 📦 v0.1.0 via 🐍 v3.13.4
❯ uv run main.py
Using CPython 3.9.6 interpreter at: /Library/Developer/CommandLineTools/usr/bin/python3
Creating virtual environment at: .venv
Installed 14 packages in 8ms
Traceback (most recent call last):
  File "/Users/sam/git/flask-admin/examples/datetime-timezone/main.py", line 9, in <module>
    from flask_admin.contrib.sqla import ModelView
  File "/Users/sam/git/flask-admin/flask_admin/contrib/sqla/__init__.py", line 2, in <module>
    from .view import ModelView
  File "/Users/sam/git/flask-admin/flask_admin/contrib/sqla/view.py", line 33, in <module>
    from flask_admin.contrib.sqla import form
  File "/Users/sam/git/flask-admin/flask_admin/contrib/sqla/form.py", line 10, in <module>
    from sqlalchemy_utils import ChoiceType
ModuleNotFoundError: No module named 'sqlalchemy_utils'
```

```suggestion
    "flask-admin[sqlalchemy-with-utils]",
```

samuelhwilliams commented on examples/auth-flask-login/pyproject.toml:
```diff
@@ -0,0 +1,13 @@
+readme = "README.md"
+requires-python = ">=3.9"
+dependencies = [
+    "flask-admin[sqlalchemy]",
```
```suggestion
    "flask-admin[sqlalchemy-with-utils]",
```

samuelhwilliams commented on examples/csp-nonce/pyproject.toml:
```diff
@@ -0,0 +1,13 @@
+readme = "README.md"
+requires-python = ">=3.9"
+dependencies = [
+    "flask-admin",
+    "flask-talisman",
```
🤔 

```
Using CPython 3.9.6 interpreter at: /Library/Developer/CommandLineTools/usr/bin/python3
Creating virtual environment at: .venv
Installed 12 packages in 5ms
Traceback (most recent call last):
  File "/Users/sam/git/flask-admin/examples/csp-nonce/main.py", line 2, in <module>
    from flask_admin import Admin
  File "/Users/sam/git/flask-admin/flask_admin/__init__.py", line 6, in <module>
    from .base import Admin  # noqa: F401
  File "/Users/sam/git/flask-admin/flask_admin/base.py", line 16, in <module>
    from flask_admin import babel
  File "/Users/sam/git/flask-admin/flask_admin/babel.py", line 67, in <module>
    from .helpers import get_current_view
  File "/Users/sam/git/flask-admin/flask_admin/helpers.py", line 20, in <module>
    from flask_admin._compat import iteritems
  File "/Users/sam/git/flask-admin/flask_admin/_compat.py", line 16, in <module>
    from flask_admin._types import T_TRANSLATABLE, T_ITER_CHOICES
  File "/Users/sam/git/flask-admin/flask_admin/_types.py", line 10, in <module>
    from typing_extensions import NotRequired
ModuleNotFoundError: No module named 'typing_extensions'
```

hasansezertasan replied:
I think we need to do some python version check when using `typing` and `typing_extensions` in the main module.

`TypeGuard` and `NotRequired` are the only two types we import from `typing_extensions`.

```python
from typing import TypeGuard  # Available on `python>=3.10`
from typing import NotRequired  # Available on `python>=3.11`
```

I think this means main module needs to depend on `typing_extension` if `python<=3.11`.

hasansezertasan replied:
#2626 should fix this problem generally. I'm resolving this comment for now.

samuelhwilliams commented on examples/custom-layout/pyproject.toml:
```diff
@@ -0,0 +1,12 @@
+readme = "README.md"
+requires-python = ">=3.9"
+dependencies = [
+    "flask-admin[sqlalchemy]",
```
```
custom-layout
Using CPython 3.9.6 interpreter at: /Library/Developer/CommandLineTools/usr/bin/python3
Creating virtual environment at: .venv
Installed 14 packages in 7ms
Traceback (most recent call last):
  File "/Users/sam/git/flask-admin/examples/custom-layout/main.py", line 6, in <module>
    from flask_admin.contrib.sqla import ModelView
  File "/Users/sam/git/flask-admin/flask_admin/contrib/sqla/__init__.py", line 2, in <module>
    from .view import ModelView
  File "/Users/sam/git/flask-admin/flask_admin/contrib/sqla/view.py", line 33, in <module>
    from flask_admin.contrib.sqla import form
  File "/Users/sam/git/flask-admin/flask_admin/contrib/sqla/form.py", line 10, in <module>
    from sqlalchemy_utils import ChoiceType
ModuleNotFoundError: No module named 'sqlalchemy_utils'
```

```suggestion
    "flask-admin[sqlalchemy-with-utils]",
```

samuelhwilliams commented on examples/s3/pyproject.toml:
```diff
@@ -0,0 +1,13 @@
+readme = "README.md"
+requires-python = ">=3.9"
+dependencies = [
+    "flask-admin[s3]",
```
```
^Cs3
Using CPython 3.9.6 interpreter at: /Library/Developer/CommandLineTools/usr/bin/python3
Creating virtual environment at: .venv
Installed 27 packages in 38ms
Traceback (most recent call last):
  File "/Users/sam/git/flask-admin/examples/s3/main.py", line 8, in <module>
    from flask_babel import Babel
ModuleNotFoundError: No module named 'flask_babel'
```

samuelhwilliams commented on examples/sqla-association_proxy/pyproject.toml:
```diff
@@ -0,0 +1,12 @@
+readme = "README.md"
+requires-python = ">=3.9"
+dependencies = [
+    "flask-admin[sqlalchemy]",
```
```suggestion
    "flask-admin[sqlalchemy-with-utils]",
```

samuelhwilliams commented on examples/sqla-custom-inline-forms/pyproject.toml:
```diff
@@ -0,0 +1,12 @@
+readme = "README.md"
+requires-python = ">=3.9"
+dependencies = [
+    "flask-admin[sqlalchemy]",
```
```suggestion
    "flask-admin[sqlalchemy-with-utils]",
```

samuelhwilliams commented on examples/geo-alchemy/README.md:
````diff
@@ -0,0 +1,38 @@
+cd flask-admin/examples/geo-alchemy
+```
+
+Setup the database:
+
+```sql
+psql postgres
+
+CREATE DATABASE flask_admin_geo;
+CREATE ROLE flask_admin_geo LOGIN PASSWORD 'flask_admin_geo';
+GRANT ALL PRIVILEGES ON DATABASE flask_admin_geo TO flask_admin_geo;
+\q
+
+psql flask_admin_geo
+
+CREATE EXTENSION postgis;
+\q
+```
````
Can we automate this as part of `main.py` and the testcontainer we use so that manually nothing needs to happen?

(We could do this as a separate PR)

hasansezertasan replied:
Let's do it in a separate PR, I want to play with the testcontainer module a bit more :nerd_face:.

samuelhwilliams reviewed (commented):
Nice work standardising all the examples, and the addition of testcontainers for some of the dependent services feels nice. Thank you! 🙇  A few of the examples don't seem to run, have tagged some, but also: geo-alchemy, host-matching, methodview, multiple-admin-instances, simple are failing on typing_extensions (for me?). Also `mongoengine`'s file isn't called main.py

hasansezertasan commented:
> Why does peewee require sqlalchemy? Haven't looked at the code yet, but that would look like a bug.


I believe we need to resolve this first.


```shell
Traceback (most recent call last):
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/peewee/main.py", line 6, in <module>
    from flask_admin.contrib.peewee import ModelView
  File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/__init__.py", line 11, in <module>
    from .view import ModelView
  File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/view.py", line 21, in <module>
    from flask_admin.contrib.peewee import filters
  File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/filters.py", line 4, in <module>
    from flask_admin.model import filters
  File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/model/__init__.py", line 2, in <module>
    from .base import BaseModelView
  File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/model/base.py", line 22, in <module>
    from sqlalchemy.orm import InstrumentedAttribute
ModuleNotFoundError: No module named 'sqlalchemy'
```

hasansezertasan commented:
I also want to remove `what comments` in the examples and leave `why comments` as is. Is it cool?

ElLorans commented:
> > Why does peewee require sqlalchemy? Haven't looked at the code yet, but that would look like a bug.
> 
> 
> 
> 
> 
> I believe we need to resolve this first.
> 
> 
> 
> 
> 
> ```shell
> 
> Traceback (most recent call last):
> 
>   File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/peewee/main.py", line 6, in <module>
> 
>     from flask_admin.contrib.peewee import ModelView
> 
>   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/__init__.py", line 11, in <module>
> 
>     from .view import ModelView
> 
>   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/view.py", line 21, in <module>
> 
>     from flask_admin.contrib.peewee import filters
> 
>   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/filters.py", line 4, in <module>
> 
>     from flask_admin.model import filters
> 
>   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/model/__init__.py", line 2, in <module>
> 
>     from .base import BaseModelView
> 
>   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/model/base.py", line 22, in <module>
> 
>     from sqlalchemy.orm import InstrumentedAttribute
> 
> ModuleNotFoundError: No module named 'sqlalchemy'
> 
> ```

Thanks. Probably fault of (my) type hints commit. If that is true, I can fix that after a separate issue and PR this weekend or you can yourself. Basically we have to add the type to _types.py and prevent an import error with TYPE_CHECKING. It is more apparent we need a test running the examples in the right environment. Proably not in this PR but a future one.

hasansezertasan commented:
> Nice work standardising all the examples, and the addition of testcontainers for some of the dependent services feels nice. Thank you! 🙇

🫶 🫶 🫶 

> A few of the examples don't seem to run, have tagged some, but also: geo-alchemy, host-matching, methodview, multiple-admin-instances, simple are failing on typing_extensions (for me?).

I have mentioned about `typing_extensions` at https://github.com/pallets-eco/flask-admin/pull/2624#discussion_r2169560682.

For the rest of the examples, they should be running properly.

>  Also `mongoengine`'s file isn't called main.py

I didn't get what you mean 🤔.

---

Additionally, I added a personal touch to the Babel Example.

samuelhwilliams commented:
Give me a ping when you're done with any final tweaks and I'll do a last review and merge :)

ElLorans commented:
> > > Why does peewee require sqlalchemy? Haven't looked at the code yet, but that would look like a bug.
> > 
> > 
> > I believe we need to resolve this first.
> > ```shell
> > Traceback (most recent call last):
> > 
> >   File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/peewee/main.py", line 6, in <module>
> > 
> >     from flask_admin.contrib.peewee import ModelView
> > 
> >   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/__init__.py", line 11, in <module>
> > 
> >     from .view import ModelView
> > 
> >   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/view.py", line 21, in <module>
> > 
> >     from flask_admin.contrib.peewee import filters
> > 
> >   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/contrib/peewee/filters.py", line 4, in <module>
> > 
> >     from flask_admin.model import filters
> > 
> >   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/model/__init__.py", line 2, in <module>
> > 
> >     from .base import BaseModelView
> > 
> >   File "/Users/hasansezertasan/Developer/projects/flask-admin/flask_admin/model/base.py", line 22, in <module>
> > 
> >     from sqlalchemy.orm import InstrumentedAttribute
> > 
> > ModuleNotFoundError: No module named 'sqlalchemy'
> > ```
> 
> Thanks. Probably fault of (my) type hints commit. If that is true, I can fix that after a separate issue and PR this weekend or you can yourself. Basically we have to add the type to _types.py and prevent an import error with TYPE_CHECKING. It is more apparent we need a test running the examples in the right environment. Proably not in this PR but a future one.

Addressed in https://github.com/pallets-eco/flask-admin/pull/2629!

ElLorans commented on examples/pymongo/pyproject.toml:
TODO: Remove sqlalchemy when https://github.com/pallets-eco/flask-admin/issues/2628 is merged

ElLorans commented:
Now that https://github.com/pallets-eco/flask-admin/issues/2628 is merged, also sqlalchemy_utils should not be required anymore, unless Choice or Currency are used.

hasansezertasan commented:
> I am converting this PR to draft. I'll re-check every example when #2626 is merged 🤓.

I couldn't check every example, except for the examples that depend on `typing_extensions`, I'm also waiting for your comments 🎊 .

@ElLorans, @samuelhwilliams 🫶.

ElLorans commented:
https://github.com/schapman1974/tinymongo looks unmantained. Does it make sense to keep the example?

ElLorans commented:
* auth-flask-login: `AttributeError: module 'hashlib' has no attribute 'scrypt'`. Are you on Mac? Maybe we can follow https://www.reddit.com/r/learnpython/comments/18lq66o/attributeerror_module_hashlib_has_no_attribute/
* flask-admin-geo (maybe needs to run a testcontainer for postgis?)
* mongoengine (maybe unsurprising): **delete example**
* sqla-association_proxy (I think this can just be deleted; we also have `sqla-association-proxy`?): **delete example**
* tinymongo loads but has errors on the model pages: **delete example**
* If it's useful/easier I think we could pin the python version for all of the examples: **I don't know.**

What is the best approach to have tests for examples? Doesn't have to be this PR.

samuelhwilliams commented:
we should open an issue to discuss e2e examples testing - i don't think useful to distract this PR :)

samuelhwilliams commented:
> I can't seem to reproduce this error (I'm on macOS) 🙄.

So am I 🤔 

```
flask-admin/examples/auth-flask-login on  chore/uv-migration/examples is 📦 v0.1.0 via 🐍 v3.13.5
❯ uv run python --version
Python 3.9.6

flask-admin/examples/auth-flask-login on  chore/uv-migration/examples is 📦 v0.1.0 via 🐍 v3.13.5
❯ uv run main.py
2025-07-25 20:40:18,195 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-07-25 20:40:18,195 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("user")
2025-07-25 20:40:18,195 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-07-25 20:40:18,195 INFO sqlalchemy.engine.Engine
DROP TABLE user
2025-07-25 20:40:18,195 INFO sqlalchemy.engine.Engine [no key 0.00003s] ()
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine COMMIT
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("user")
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("user")
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine [raw sql] ()
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine
CREATE TABLE user (
        id INTEGER NOT NULL,
        first_name VARCHAR(100),
        last_name VARCHAR(100),
        login VARCHAR(80),
        email VARCHAR(120),
        password VARCHAR(64),
        PRIMARY KEY (id),
        UNIQUE (login)
)


2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine [no key 0.00002s] ()
2025-07-25 20:40:18,196 INFO sqlalchemy.engine.Engine COMMIT
Traceback (most recent call last):
  File "/Users/sam/git/flask-admin/examples/auth-flask-login/main.py", line 269, in <module>
    build_sample_db()
  File "/Users/sam/git/flask-admin/examples/auth-flask-login/main.py", line 175, in build_sample_db
    test_user = User(login="test", password=generate_password_hash("test"))
  File "/Users/sam/git/flask-admin/examples/auth-flask-login/.venv/lib/python3.9/site-packages/werkzeug/security.py", line 108, in generate_password_hash
    h, actual_method = _hash_internal(method, salt, password)
  File "/Users/sam/git/flask-admin/examples/auth-flask-login/.venv/lib/python3.9/site-packages/werkzeug/security.py", line 43, in _hash_internal
    hashlib.scrypt(
AttributeError: module 'hashlib' has no attribute 'scrypt'
```

> If you are talking about pymongo example, It just works 😕.

> I renamed sqla-association_proxy as sqla-association-proxy. Should I just delete the renamed version ❔.

Apologies - I still had old copies of these dirs floating around 🤦 

> Edit: GeoAlchemy example requires some pre-installed software to work. I'm on a mac and I had to do brew install postgresqlto make it work. I added this step to the instructions section

I'll look at this quickly - if it just needs a postgres DB running, I think we can testcontainers to remove that manual setup/dependency.

hasansezertasan commented:
> * auth-flask-login: `AttributeError: module 'hashlib' has no attribute 'scrypt'`. Are you on Mac? Maybe we can follow [reddit.com/r/learnpython/comments/18lq66o/attributeerror_module_hashlib_has_no_attribute](https://www.reddit.com/r/learnpython/comments/18lq66o/attributeerror_module_hashlib_has_no_attribute/)

Thanks to @ElLorans, I was able to fix that `scrypt` error 😎 😇.

hasansezertasan commented:
> Apologies - I still had old copies of these dirs floating around 🤦

That's OK 🙏.

> I'll look at this quickly - if it just needs a postgres DB running, I think we can testcontainers to remove that manual setup/dependency.

OK, I'll be waiting to hear from you.

> edit: why exactly did you need to install postgres? It's already using a testcontainer 🤔 And its testcontainer has your username as username+password - can we make those generic please? username=`postgres`, password=`password` is probably fine

I just updated those, sorry. I forgot 😜.

---

At first, I got an error similar to this: https://stackoverflow.com/questions/33866695/error-installing-psycopg2-on-macos-10-9-5

Using `psycopg2-binary`didn't work out but https://stackoverflow.com/a/33866865/11065393 did.

I just removed the `postgresql`from my system and tried to run the example again, here is my output:

```sh
Using CPython 3.9.23 interpreter at: /Users/hasansezertasan/.local/share/mise/installs/python/3.9.23/bin/python3.9
Creating virtual environment at: .venv
Installed 37 packages in 44ms
Traceback (most recent call last):
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/main.py", line 88, in <module>
    db.init_app(app)
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/flask_sqlalchemy/extension.py", line 374, in init_app
    engines[key] = self._make_engine(key, options, app)
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/flask_sqlalchemy/extension.py", line 665, in _make_engine
    return sa.engine_from_config(options, prefix="")
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/sqlalchemy/engine/create.py", line 823, in engine_from_config
    return create_engine(url, **options)
  File "<string>", line 2, in create_engine
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/sqlalchemy/util/deprecations.py", line 281, in warned
    return fn(*args, **kwargs)  # type: ignore[no-any-return]
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/sqlalchemy/engine/create.py", line 602, in create_engine
    dbapi = dbapi_meth(**dbapi_args)
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 696, in import_dbapi
    import psycopg2
  File "/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/psycopg2/__init__.py", line 51, in <module>
    from psycopg2._psycopg import (                     # noqa
ImportError: dlopen(/Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/psycopg2/_psycopg.cpython-39-darwin.so, 0x0002): Library not loaded: /opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.5.dylib
  Referenced from: <F5E604FD-604B-36A0-A56B-31B232F89530> /Users/hasansezertasan/Developer/projects/flask-admin/examples/geo-alchemy/.venv/lib/python3.9/site-packages/psycopg2/_psycopg.cpython-39-darwin.so
  Reason: tried: '/opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.5.dylib' (no such file), '/opt/homebrew/opt/postgresql@14/lib/postgresql@14/libpq.5.dylib' (no such file)
```

hasansezertasan commented:
> Mypy crashes on folders named with '-', I think. So I would move to '_'.

Are we running mypy on examples 😕?

hasansezertasan commented:
> No, but we could

I see, should we do it then ❔

hasansezertasan commented:
> I do not get any errors, but I cannot see the map below the points/shapes I create. That looks because we need a FLASK_ADMIN_MAPBOX_ACCESS_TOKEN, but can't we specify it graphically in the example, too?

I think it's possible through `current_app.config`.

ElLorans commented:
> > I do not get any errors, but I cannot see the map below the points/shapes I create. That looks because we need a FLASK_ADMIN_MAPBOX_ACCESS_TOKEN, but can't we specify it graphically in the example, too?
> 
> I think it's possible through `current_app.config`.

let me explain better: when the user runs the example, does it make sense to add a graphical message stating he needs the API key? On the one hand I think it's trivial, on the other for a millisecond I thought the example was not working

hasansezertasan commented:
> let me explain better: when the user runs the example, does it make sense to add a graphical message stating he needs the API key? On the one hand I think it's trivial, on the other for a millisecond I thought the example was not working

Well, there are READMEs for each example and the README.md file in the `geo-alchemy` already states why the maps are not rendered and how to make them work 🤓.

Rather than adding a graphical message to this example, it would be better to show some kind of indicator where the maps should be, stating that there are missing credentials to make the feature work 😎. And I think it should be a part of Flask Admin itself.

ElLorans commented:
Makes sense. Then I think all examples are now working and we can merge.

hasansezertasan commented:
> Makes sense. Then I think all examples are now working and we can merge.

Fantastic 🥳. I just need to check something before going forward, ripgrep my name in the repository 🤓.

...

All is safe 🫶.

samuelhwilliams commented:
> I'll look at this quickly - if it just needs a postgres DB running, I think we can testcontainers to remove that manual setup/dependency.

Cheers for explaining - the postgres local dependency makes sense with psycopg2/etc, so all good :)

I'll do another quick scan but I think this is all good now - thanks so much for doing the work 🙇
